moq-karp/src/broadcast.rs

```diff
@@ -1,116 +1,175 @@
-use crate::{Audio, Catalog, Error, Result, Track, TrackConsumer, TrackProducer, Video};
+use crate::{Audio, Catalog, Error, Result, Track, TrackProducer, Video};
 
 use moq_async::{spawn, Lock};
 use moq_transfork::{Announced, AnnouncedConsumer, AnnouncedMatch, Session};
+use crate::track::TrackConsumer;
 
 use derive_more::Debug;
 
+#[derive(Clone)]
+struct BroadcastListener {
+	pub session: Session,
+	pub catalog: Lock<CatalogProducer>,
+}
+
+#[derive(Debug, Clone)]
+struct Context {
+	pub video: Option<Video>,
+	pub audio: Option<Audio>,
+	pub video_consumer: Option<moq_transfork::TrackConsumer>,
+	pub audio_consumer: Option<moq_transfork::TrackConsumer>,
+}
+impl Context {
+	pub fn default() -> Self {
+		Self {
+			video: None,
+			audio: None,
+			video_consumer: None,
+			audio_consumer: None,
+		}
+	}
+
+	fn publish_video_to(&self, listener: &mut BroadcastListener) -> anyhow::Result<()> {
+		if let Some(consumer) = self.video_consumer.clone() {
+			if let Some(info) = self.video.clone() {
+				let mut catalog = listener.catalog.lock();
+				catalog.current.video.push(info.clone());
+				catalog.publish()?;
+
+				// let consumer = producer.subscribe();
+				let consumer = TrackConsumer::new(consumer);
+				listener.session.publish(consumer.track.clone())?;
+
+				// Start a task that will remove the catalog on drop.
+				let catalog = listener.catalog.clone();
+				let track_info = info.track;
+				spawn(async move {
+					consumer.closed().await.ok();
+
+					let mut catalog = catalog.lock();
+					catalog.current.video.retain(|v| v.track != track_info);
+					catalog.publish().unwrap();
+				});
+			}
+		}
+
+		Ok(())
+	}
+
+	pub fn publish_audio_to(&self, listener: &mut BroadcastListener) -> anyhow::Result<()> {
+		if let Some(consumer) = self.audio_consumer.clone() {
+			if let Some(info) = self.audio.clone() {
+				let mut catalog = listener.catalog.lock();
+				catalog.current.audio.push(info.clone());
+				catalog.publish()?;
+
+				// let consumer = producer.subscribe();
+				let consumer = TrackConsumer::new(consumer);
+				listener.session.publish(consumer.track.clone())?;
+
+				// Start a task that will remove the catalog on drop.
+				let catalog = listener.catalog.clone();
+				let track_info = info.track;
+				spawn(async move {
+					consumer.closed().await.ok();
+
+					let mut catalog = catalog.lock();
+					catalog.current.audio.retain(|v| v.track != track_info);
+					catalog.publish().unwrap();
+				});
+			}
+		}
+
+		Ok(())
+	}
+}
+
 #[derive(Debug, Clone)]
```

Comment:
Yeah this `Clone` is why you had to remove `BroadcastProducer` from `Import`. That's because the listeners `Vec` would not be shared between the multiple handles. If you put `listeners` behind a `Lock` then it should just work.

Reply:
We could remove the `Clone` but someone wanted a way to get access to the current catalog. I think we keep it for now.